runtime/web_worker.rs

```diff
@@ -136,7 +139,21 @@ impl WebWorkerInternalHandle {
 
   /// Check if this worker is terminated or being terminated
   pub fn is_terminated(&self) -> bool {
-    self.terminated.load(Ordering::SeqCst)
+    self.has_terminated.load(Ordering::SeqCst)
+  }
+
+  /// Check if this worker must terminate (because the termination signal is
+  /// set), and terminates it if so. Returns whether the worker is terminated or
+  /// being terminated, as with [`Self::is_terminated()`].
+  pub fn terminate_if_needed(&mut self) -> bool {
+    let has_terminated = self.is_terminated();
+
+    if !has_terminated && self.termination_signal.load(Ordering::SeqCst) {
+      self.terminate();
+      return true;
+    }
+
+    has_terminated
   }
```

Comment:
This method can end up calling `self.terminate()` repeatedly, can't it? That seems bad.

Reply:
`terminate_if_needed()` will at most call `self.terminate()` once – there are no loops or recursion in this method. And if `self.terminate()` has been previously called, `self.is_terminated()` will return true, so `self.terminate()` won't be called again.

Reply:
Sorry, let me clarify: calling `terminate_if_needed()` twice can result in two calls to `terminate()`.

That's okay when `terminate()` is idempotent but I don't think it is, and even if it is, then that's an implicit contract that's easy to miss when making changes later

Reply:
`self.terminate()` will set `self.has_terminated` to true, which will keep any subsequent call to `self.terminate_if_needed()` from calling `self.terminate()` again.

Reply:
That's the implicit contract I mean. Something a little less prone to snafus would be appreciated.

A rule of thumb I follow is this: whenever more than one atomic variable pops up, try really hard to replace it with a mutex and a proper critical section.